docs/Style.thy

```diff
@@ -0,0 +1,182 @@
+(*
+ * Copyright 2021, Data61, CSIRO
+ *
+ * This software may be distributed and modified according to the terms of
+ * the BSD 2-Clause license. Note that NO WARRANTY is provided.
+ * See "LICENSE_BSD2.txt" for details.
+ *
+ * @TAG(DATA61_BSD)
+ *)
+
+chapter \<open>An Isabelle Syntax Style Guide\<close>
+
+theory Style
+  imports Main
+begin
+
+(* This style guide is heavily influenced by Gerwin's Style Guide for Isabelle/HOL. See parts 1
+   and 2 here:
+
+   * https://proofcraft.org/blog/isabelle-style.html
+   * https://proofcraft.org/blog/isabelle-style-part2.html
+
+   This guide may repeat some parts of Gerwin's guide and should not conflict in any way.
+ *)
+
+section \<open>General Principles\<close>
+
+(* TO DO: What advice should go here? *)
+
+
+section \<open>Definitions\<close>
+
+(* Definitions should be written over 2 lines in the following style.
+
+   Note that indentation is 2 spaces and the definitions do not run past 100 character line
+   limit. Note also that the \equiv symbol is used.
+ *)
+
+definition hd_opt :: "'a list \<Rightarrow> 'a option" where
+  "hd_opt \<equiv> case_list None (\<lambda>h t. Some h)"
+
+(* In the case that the type_signature causes an overflow of the 100 char line limit,
+   use the following pattern.
+
+   Note that the indentation of the type signature remains no less than the indentation level of
+   the initial `"`.
+ *)
+
+type_synonym my_very_long_nat_type = nat
+type_synonym 'a my_very_long_option_type = "'a option"
+type_synonym 'a my_very_long_list_type = "'a list"
+definition hd_opt2 :: "'a my_very_long_list_type \<Rightarrow> my_very_long_nat_type \<Rightarrow>
+                       'a my_very_long_list_type \<Rightarrow> 'a my_very_long_option_type" where
```

Comment:
I would not indent this way, actually. I'd write

```
definition hd_opt2 ::
  "'a my_very_long_list_type \<Rightarrow> my_very_long_nat_type \<Rightarrow>
   'a my_very_long_list_type \<Rightarrow> 'a my_very_long_option_type"
  where
```

`where` could also go on the end of the last line, I don't mind that. The main point is to break after the `::` to avoid a potentially very long right overhang (length depends on the name of the constant).